Complex Types client library

mregen opened the pull request:
see also #838 

- Adds a complex type system client library extension for custom enumerations and structure types based on a server dictionary or DataTypeDefinition using emitted classes at runtime. 
- Custom types are added to the session factory, not to the global factory. Types are loaded when the session connects by adding 
`                var complexTypeSystem = new ComplexTypeSystem(session);
                await complexTypeSystem.Load();
`

- Custom type properties can be accessed with an accessor [propertyname] or [index].
- Support includes V1.03 dictionaries and V1.04 DataTypeDefinitions (--> known issues).
- NetCoreComplexClient can be used for quick testing servers. It prints out all custom types with content and subscribes to all. Also includes sample how to write back complex type variables.
- The UA Sample Client and the Quickstart Data Type client can be used for testing with UI.
- Caveat: The complex type libaries for .Net Core must be built for .netcoreapp, so it needs a seperate nuget package for the complex type client extension.

Finding enough servers with custom types is an issue, so if you find servers with unsupported custom types please provide feedback.

Known issues:
- UA Sample Client hits still plenty of UI issues when writing custom types. UI Controls are not aware of unions, optional fields and option set.
- Only V1.03 binary dictionaries are supported (no XML dictionaries).
- V1.03 dictionary structure definitions are only supported if they can be mapped to a V1.04 DataTypeDefinition. Nested types and cross dictionary refs should work.
- OptionSet is only partially implemented. Only shows the byte arrays, not the flags.
- Opaque Types are ignored.

mregen commented on SampleApplications/SDK/Opc.Ua.Client.ComplexTypes/Opc.Ua.Client.ComplexTypes.csproj:
```diff
@@ -0,0 +1,30 @@
+    <TargetFrameworks>net46;netcoreapp2.0</TargetFrameworks>
+    <AssemblyName>Opc.Ua.Client.Complex</AssemblyName>
+    <LangVersion>7</LangVersion>
+    <PackageId>Opc.Ua.Client.Complex</PackageId>
```
>Opc.Ua.Client.Complex [](start = 15, length = 21)

package / type name

marcschier commented on SampleApplications/SDK/Opc.Ua.Client.ComplexTypes/BaseComplexType.cs:
```diff
@@ -0,0 +1,923 @@
+        private void EncodeProperty(IEncoder encoder, PropertyInfo property)
+        {
+            var propertyType = property.PropertyType;
+            if (propertyType == typeof(Boolean))
```
Minor: It would be great to have this as extension methods for IEncoder, i.e. Write<T>().      I have seen the same if/else in pub/sub.   The problem is that when we add more types (e.g. decimal), a lot of code needs to be touched.

mregen replied:
also for later ...

marcschier commented on SampleApplications/SDK/Opc.Ua.Client.ComplexTypes/ComplexTypeSystem.cs:
```diff
@@ -0,0 +1,792 @@
+        /// Load all subTypes and optionally nested subtypes of a type definition.
+        /// Filter for all subtypes or only subtypesoutside the default namespace.
+        /// </summary>
+        private IList<INode> LoadDataTypes(ExpandedNodeId dataType, bool nestedSubTypes = false, bool filterUATypes = true)
```
Would be cool to have an async or Begin/End version of it also - eventually....

mregen replied:
It is private so it can be improved later...

mregen commented:
few updates:
- Added fixes for UA Sample Client. 
- Added factory class to allow to plugin other type builder methods (e.g. Codedom etc...)
- Added JSON test case and include the complex types client in the testing shell script, so the complex types lib is tested on mac/linux too.
- Few fixes in the JSON encoders.
- include feedback
